protecode/util.py

```diff
@@ -94,9 +93,15 @@ def _upload_tasks():
         for component in components:
             component_groups[component.name].append(component)
 
-        def group_resources(components):
+        def group_by_resource_name(
+            components
+        ) -> typing.Generator[
+            list[cnudie.util.ComponentResource],
+            None,
+            None
+        ]:
             # groups resources of components by resource name
-            resource_groups = collections.defaultdict(list)
+            resource_groups = collections.defaultdict(list[cnudie.util.ComponentResource])
```

Comment:
interesting. I would not have guessed that this works (considering the argument of defaultdict is used to construct new instances for absent keys). Seems like a good pattern - did you come up w/ that?

Reply:
well I just added the type hint and it worked 😅